### PR TITLE
RFC: add new var system

### PR DESCRIPTION
**Pull request by jsquyres**
Stop the proliferation of unstandardized environment variables and use a standardized convention.

The first commit adds the var system: `fi_var_register()`, `fi_var_get()`, and `fi_var_fini()`.

The second commit converts the sockets provider to use the var interface:
* SOCK_PE_WAITTIME -> FI_SOCKETS_PE_WAITTIME
* SOCK_DGRAM_DROP_RATE -> FI_SOCKETS_DGRAM_DROP_RATE

@jithinjosepkl @shantonu please review

The third commit converts the PSM provider to use the var interface (note: I cannot compile/test this commit, and I'm quite sure that my help strings are totally wrong):
* OFI_PSM_NAME_SERVER -> FI_PSM_NAME_SERVER
* OFI_PSM_AM_MSG -> FI_PSM_AM_MSG
* OFI_PSM_TAGGED_RMA -> FI_PSM_TAGGED_RMA
* OFI_PSM_UUID -> FI_PSM_UUID

@j-xiong please review

### CODE REVIEW

**Comment by jsquyres** [comment]
Note that #177 is the original issue from which this PR emerged.

**Comment by shefty** [comment]
Note that Jianxin is out until Thursday of this week.

File: prov/psm/src/psmx_init.c [hunk 1]
```diff
@@ -450,11 +450,25 @@ struct fi_provider psmx_prov = {
...
+			char **value)
+{
+	*value = NULL;
+
+	if (fi_var_register(&psmx_prov, name, help_string))
+		FI_INFO(&psmx_prov, FI_LOG_CORE,
+			"Failed to register name variable\n", name);
+	if (fi_var_get(&psmx_prov, name, value))
+		FI_INFO(&psmx_prov, FI_LOG_CORE,
+			"Failed to read uuid variable\n");
```
Comment (shefty) [inline prov/psm/src/psmx_init.c]: We should be able to avoid these FI_INFO calls in each provider by moving them into the fi_var_* calls.

File: src/var.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+	v->next = fi_vars;
+	fi_vars = v;
+
+	asprintf(&str, "registered var %s\n", var_name);
+	FI_INFO(provider, FI_LOG_CORE, str);
+	free(str);
+
+	return FI_SUCCESS;
+}
```
Comment (shefty) [inline src/var.c]: Please return negative errno values in error cases.

File: src/var.c [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+				asprintf(&str, "read var %s=%s\n",
+					var_name, *value);
+			else
+				asprintf(&str, "read var %s=<not set>\n",
+					var_name);
+			FI_INFO(provider, FI_LOG_CORE, str);
+			free(str);
+
+			return FI_SUCCESS;
+		}
```
Comment (shefty) [inline src/var.c]: nit - can you add braces to the for loop?  I get lost with multi-line code segments

File: include/rdma/fi_var.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * Upon successful return, FI_SUCCESS is returned and (*value) points
+ * to a \0-terminated string value, or NULL if it was not set.  The
+ * caller should not modify or free (*value).
+ *
+ * FI_ENOENT will be returned if var_name was not previously
+ * registered via fi_var_register().
+ */
+int fi_var_get(struct fi_provider *provider, const char *var_name,
+	char **value);
```
Comment (shefty) [inline include/rdma/fi_var.h]: Does it make sense to provide helper wrappers that convert value to the correct value?  E.g.

int fi_var_get_int(*provider, *var_name, *default_value);

File: prov/sockets/src/sock_fabric.c [hunk 5]
```diff
@@ -580,19 +581,29 @@ SOCKETS_INI
 {
 	char *value;
 
-	if ((value = getenv("SOCK_PE_WAITTIME"))) {
+	if (fi_var_register(&sock_prov, "pe_waittime",
+				"How many miliseconds to spin while waiting for progress"))
+		SOCK_LOG_INFO("Failed to register pe_waittime variable\n");
+	if (fi_var_get(&sock_prov, "pe_waittime", &value))
+		SOCK_LOG_INFO("Failed to read pe_waittime variable\n");
```
Comment (shantonu) [inline prov/sockets/src/sock_fabric.c]: Since the var is now FI_SOCKETS_*, should the log report the same? The name of the var is not consistent in the logs.

File: prov/sockets/src/sock_fabric.c [hunk 6]
```diff
@@ -580,19 +581,29 @@ SOCKETS_INI
...
-		if(sock_dgram_drop_rate)
-			SOCK_LOG_INFO("SOCK_DGRAM_DROP_RATE = %d\n", sock_dgram_drop_rate);
+	if (fi_var_register(&sock_prov, "dgram_drop_rate",
+				"Drop every Nth dgram frame (debug only)"))
+		SOCK_LOG_INFO("Failed to register dgram_drop_rate variable\n");
+	if (fi_var_get(&sock_prov, "dgram_drop_rate", &value))
+		SOCK_LOG_INFO("Failed to read dgram_drop_rate variable\n");
+	if (value) {
+		sock_dgram_drop_rate = strtol(value, NULL, 10);
+		SOCK_LOG_INFO("SOCK_DGRAM_DROP_RATE = %d\n",
```
Comment (shantonu) [inline prov/sockets/src/sock_fabric.c]: Same here

File: prov/psm/src/psmx_init.c (hunk 1 above)
Comment (jsquyres) [inline prov/psm/src/psmx_init.c]: Sure; done.

File: src/var.c (hunk 2 above)
Comment (jsquyres) [inline src/var.c]: K.

File: src/var.c (hunk 3 above)
Comment (jsquyres) [inline src/var.c]: Sure.  I actually like braces even for 1-line blocks (defensive programming and all that).  But I thought you kernel guys were allergic to braces, so I left them off whenever possible.  I'm happy to add them back.  :-)

File: include/rdma/fi_var.h (hunk 4 above)
Comment (jsquyres) [inline include/rdma/fi_var.h]: Mmm.  Good point.  I don't have an opinion here -- I'm happy to add the following:

* `_get_str()`
* `_get_int()`
* `_get_long()` 
* `_get_bool()` (which takes 0, true, yes, on, 1, false, no, off -- case insensitive, of course)

That would cover all the cases in sockets and psm.

Want me to add these?

File: prov/sockets/src/sock_fabric.c (hunk 5 above)
Comment (jsquyres) [inline prov/sockets/src/sock_fabric.c]: Fair enough -- there's actually a subtle distinction between the FI configuration variable name and how it is exposed.  Currently, FI config variables are exposed in the environment via the form $FI_(upper_case(provider_name))_(upper_case(variable_name)).  E.g.:

```
shell$ export FI_SOCKETS_PE_WAITTIME=10
```

But I would expect that if we supported config files someday, the variable name would be exposed via "(prover_name)_(variable_name)", e.g., perhaps:

```ini
sockets_pe_waittime = 10
```

(and perhaps the name in the config file would be case insensitive...?)

So there's two things different:

1. The `FI_` prefix in the env var case
1. The upper / lower case issues

I feel strongly that we need the first one (`FI_` prefix), but I could be talked out of the second (upper/lower case transmogrification).

File: src/var.c [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	for (v = fi_vars; v; v = v->next)
+		if (strcmp(v->provider->name, provider->name) == 0 &&
+			strcmp(v->var_name, var_name) == 0) {
+			*value = getenv(v->env_var_name);
+
+			if (*value)
+				asprintf(&str, "read var %s=%s\n",
+					var_name, *value);
+			else
+				asprintf(&str, "read var %s=<not set>\n",
```
Comment (pmmccorm) [inline src/var.c]: Can we avoid the malloc every iteration with asprintf by doing something like:

    FI_INFO(provider, FI_LOG_CORE, "read var %s=%s\n", var_name, value);

?

**Comment by jsquyres** [comment]
I just pushed some new commits, which hid some of the prior comments by default on the PR.  I think there's two main open questions left:

1. @shefty Whether we want more `fi_var_get_TYPE()` functions (https://github.com/ofiwg/libfabric/pull/1055#discussion_r31969918)
1. @shantonu Whether we like the dichotomy of the registered variable name vs. how it appears in the environment and/or log messages (https://github.com/ofiwg/libfabric/pull/1055#discussion-diff-32030642)

File: include/rdma/fi_var.h (hunk 4 above)
Comment (shefty) [inline include/rdma/fi_var.h]: Yes, please.  Thanks

File: prov/sockets/src/sock_fabric.c (hunk 5 above)
Comment (shefty) [inline prov/sockets/src/sock_fabric.c]: I agree with keeping FI_.  I would go with case insensitive.

File: prov/psm/src/psmx_init.c [hunk 8]
```diff
@@ -475,10 +485,19 @@ PSM_INI
...
+				1);
+	psmx_env.am_msg =
+		psmx_get_int_env("am_msg",
+				"Whether PSM supports active messages or not",
+				0);
+	psmx_env.tagged_rma =
+		psmx_get_int_env("tagged_rma",
+				"Whether PSM supports tagged RMA or not",
+				1);
+	psmx_reg_and_read("uuid", "Unique ID for PSM fabric", &psmx_env.uuid);
```
Comment (shefty) [inline prov/psm/src/psmx_init.c]: @j-xiong - This is separate from this patch, but can you look at moving reading these values until immediately before they are needed?

File: prov/sockets/src/sock_fabric.c [hunk 9]
```diff
@@ -580,20 +581,21 @@ SOCKETS_INI
...
-	if(getenv("SOCK_DGRAM_DROP_RATE") != NULL) {
-		sock_dgram_drop_rate =  strtol(getenv("SOCK_DGRAM_DROP_RATE"), NULL, 10);
-		if(sock_dgram_drop_rate)
-			SOCK_LOG_INFO("SOCK_DGRAM_DROP_RATE = %d\n", sock_dgram_drop_rate);
-	}
+	fi_var_register(&sock_prov, "dgram_drop_rate",
+			"Drop every Nth dgram frame (debug only)");
+	fi_var_get(&sock_prov, "dgram_drop_rate", &value);
+	if (value)
+		sock_dgram_drop_rate = strtol(value, NULL, 10);
```
Comment (shefty) [inline prov/sockets/src/sock_fabric.c]: @jithinjosepkl - This is separate from this patch, but can you look at moving reading these values until immediately before they are needed?

File: src/var.c [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		return -FI_ENOMEM;
+	}
+
+	for (i = 0; v->env_var_name[i]; ++i)
+		v->env_var_name[i] = toupper(v->env_var_name[i]);
+
+	v->next = fi_vars;
+	fi_vars = v;
+
+	asprintf(&str, "registered var %s\n", var_name);
```
Comment (pmmccorm) [inline src/var.c]: Ditto here about asprintf.

File: prov/sockets/src/sock_fabric.c (hunk 9 above)
Comment (jithinjosepkl) [inline prov/sockets/src/sock_fabric.c]: Sure @shefty, makes sense.

File: src/var.c (hunk 7 above)
Comment (jsquyres) [inline src/var.c]: I missed that FI_INFO was varargs.  Sure, that's a whole lot better -- I'll update here and the other location.

File: prov/psm/src/psmx_init.c (hunk 8 above)
Comment (j-xiong) [inline prov/psm/src/psmx_init.c]: Yes, will do that in a separate PR.

File: prov/psm/src/psmx_init.c [hunk 11]
```diff
@@ -475,10 +485,19 @@ PSM_INI
 	int check_version;
 	int err;
 
-	psmx_env.name_server	= psmx_get_int_env("OFI_PSM_NAME_SERVER", 1);
-	psmx_env.am_msg		= psmx_get_int_env("OFI_PSM_AM_MSG", 0);
-	psmx_env.tagged_rma	= psmx_get_int_env("OFI_PSM_TAGGED_RMA", 1);
-	psmx_env.uuid		= getenv("OFI_PSM_UUID");
+	psmx_env.name_server =
+		psmx_get_int_env("name_server",
+				"Whether there is a PSM name server or not",
```
Comment (j-xiong) [inline prov/psm/src/psmx_init.c]: "Whether to turn on the name server or not"

File: prov/psm/src/psmx_init.c [hunk 12]
```diff
@@ -475,10 +485,19 @@ PSM_INI
...
+				1);
+	psmx_env.am_msg =
+		psmx_get_int_env("am_msg",
+				"Whether PSM supports active messages or not",
```
Comment (j-xiong) [inline prov/psm/src/psmx_init.c]: "Whether to use active message based messaging or not (default no)"

File: prov/psm/src/psmx_init.c [hunk 13]
```diff
@@ -475,10 +485,19 @@ PSM_INI
...
+				0);
+	psmx_env.tagged_rma =
+		psmx_get_int_env("tagged_rma",
+				"Whether PSM supports tagged RMA or not",
```
Comment (j-xiong) [inline prov/psm/src/psmx_init.c]: "Whether to use tagged message for large size RMA or not (default yes)"

File: prov/psm/src/psmx_init.c (hunk 8 above)
Comment (j-xiong) [inline prov/psm/src/psmx_init.c]: "Unique Job ID required by PSM fabric"

File: prov/psm/src/psmx_init.c (hunk 11 above)
Comment (jsquyres) [inline prov/psm/src/psmx_init.c]: Fixed.

File: prov/psm/src/psmx_init.c (hunk 12 above)
Comment (jsquyres) [inline prov/psm/src/psmx_init.c]: Fixed.

File: prov/psm/src/psmx_init.c (hunk 13 above)
Comment (jsquyres) [inline prov/psm/src/psmx_init.c]: Fixed.

File: prov/psm/src/psmx_init.c (hunk 8 above)
Comment (jsquyres) [inline prov/psm/src/psmx_init.c]: Fixed.

**Comment by jsquyres** [comment]
New commits pushed:

* Updated var interface to have "get" methods for string, int, long, and boolean
* Updated psm and sockets providers to match (I still can't compile/test the PSM provider, sorry!)

Comments on the new proposed API welcome (this is an RFC, after all).  See fi_var.h for a description of the API functions.

File: src/var.c [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+					"read boolean var %s=false\n", var_name);
+				return FI_SUCCESS;
+			} else if (strcmp(var_name, "1") == 0 ||
+				strcasecmp(var_name, "true") == 0 ||
+				strcasecmp(var_name, "yes") == 0 ||
+				strcasecmp(var_name, "on") == 0) {
+				*value = 1;
+				FI_INFO(provider, FI_LOG_CORE,
+					"read boolean var %s=true\n", var_name);
+				return -FI_ENODATA;
```
Comment (shefty) [inline src/var.c]: Should return FI_SUCCESS here.

File: src/var.c [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+	if (ret == FI_SUCCESS) {
+		if (*value) {
+			FI_INFO(provider, FI_LOG_CORE,
+				"read string var %s=%s\n", var_name, *value);
+			return FI_SUCCESS;
+		} else {
+			FI_INFO(provider, FI_LOG_CORE,
+				"read string var %s=<not set>\n", var_name);
+			return -FI_ENODATA;
+		}
```
Comment (shefty) [inline src/var.c]: You can drop the 'return FI_SUCCESS' line and just let the code fall through.  And change 'return -FI_ENODATA' to 'ret = -FI_ENODATA', which will give us a single exit point.  This same flow can be followed below.

**Comment by shefty** [comment]
I'm pretty happy with this.

File: include/rdma/fi_var.h [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+ * No checking is done to ensure that the value the user set is
+ * actually an integer -- atoi() is simply called on whatever value
+ * the user sets.
+ */
+int fi_var_get_int(struct fi_provider *provider, const char *var_name,
+	int *value);
+
+/* Similar to fi_var_get_str(), but the value is converted to an
+ * long.  No checking is done to ensure that the value the user set
+ * is actually an integer -- atoi() is simply called on whatever value
```
Comment (shefty) [inline include/rdma/fi_var.h]: actually uses strtol(), rather than atoi().

File: src/var.c (hunk 14 above)
Comment (jsquyres) [inline src/var.c]: Yoinks; thanks.

**Comment by jsquyres** [comment]
New commits pushed.

Would like to get signoffs from the sockets and PSM providers before merging (I cannot test the PSM stuff).

**Comment by jithinjosepkl** [comment]
:+1: from sockets. But, I was just wondering where we use the help strings.
Is there a way user can list all the options for a particular provider?
(May be this is something thats already in plan?)

**Comment by jsquyres** [comment]
Ja, the plan for the help strings is to output them in fi_info, or somesuch.  Not sure yet -- but there will be some kind of CLI utility to display them for the user.
